proposed/2022/cpm-dotnet-cli-add-package-support.md

```diff
@@ -1,29 +1,46 @@
+Projects onboarded to CPM use a `Directory.packages.props` file in the root of the repo where package versions are defined centrally. Ideally, when the `dotnet add package` command is used, the package version should only be added to the corresponding package in the `Directory.packages.props` file. However, currently the command attempts to add the package version to the `<PackageReference />` in the project which conflicts with the CPM requirements that package versions must only be in the `Directory.packages.props` file.
 
 ## Goals
+
 The main goal is to add support for `dotnet add package` to be used with projects onboarded onto CPM. Regardless of whether the package has already been added to the project or not, the command should allow users to add packages or update the package version in the `Directory.packages.props` file.
 
 ## Customers
+
 Users wanting to use CPM onboarded projects and dotnet CLI commands.
 
-## Solutions
+## Solution
+
 When `dotnet add package` is executed in a project onboarded to CPM (meaning that the `Directory.packages.props` file exists) there are a few scenarios that must be considered.
 
+| Scenario # | PackageReference exists? | Directory.Packages.Props file exists? | Is ManagePackageVersionsCentrally property set to true? | Is VersionOverride? | Current behavior | New behavior in dotnet CLI | In Scope |
+| ---- |-----| -----|-----|----------|----------|----------|----------|
+| 1 | ❌ | ✔️ | ✔️ (in Directory.Packages.Props or .csproj file) | ❌ | Restore failed with NU1008 error and NO edits were made to the csproj file (same in VS and dotnet CLI) | `PackageReference` should be added to .(cs/vb)proj file and `PackageVersion` should be added to the closest `Directory.Packages.Props` file | ✔️ |
+| 2 | ❌ | ❌ | ✔️ in (cs/vb)proj file | ❌ | Restore failed with NU1008 error and NO edits were made to the csproj file (same in VS and dotnet CLI) | `PackageReference` and `Version` should be added to .(cs/vb)proj file.| ✔️ |
+| 3 | ❌ | ✔️ | ❌ | ❌ | `PackageReference` and `Version` added to .(cs/vb)proj file. | In addition to the current behavior `Directory.packages.props` file should be deleted if it exists in the project folder | ❌ |
+| 4 | ✔️ | ✔️ | ✔️ | ❌ | **dotnet CLI** - Restore failed with NU1008 error and NO edits were made to the csproj file. **VS** - Clicked on Update package in PM UI. New `PackageVersion` added to csproj file and `Directory.packages.props` file was not updated | No changes should be made to the `PackageReference` item in .(cs/vb)proj file but `PackageVersion` item should be updated with the new version in the appropriate `Directory.Packages.Props` file.|✔️ [More Info](https://github.com/NuGet/Home/pull/11849#discussion_r890639808) |
```

Comment:
I don't think add package should update versions. 
This is something that's not readily apparent to the user. 

Imo, the behavior should be either version override or an explciit command to update in CPM depending on whether --version has been passed it or not.

Reply:
We went down on this path because `dotnet add package` currently updates the package version if the reference already exists. In the case of CPM, updating version in the `Directory.Packages.Props` file is an option.

Your suggestion to add `VersionOverride` if the package reference already exists is also a good one. @jeffkl what do you think?

Reply:
> In my opinion, its just more work for a user to use a different gesture for add vs update. They might be running it on a bunch of projects via automation and they'd have to determine whether not that package is already referenced. So I think add package should just update the version for them.

The reason why I bring it up is because we have issues and customer interviews that have suggested otherwise. 
There's also an update for a `dotnet update package` command, which is something we'd want to do eventually as well. 

I think dotnet add package should match the VS experience for install package.

When you go to an individual project and run `add package`, you don't expect the package to be updated for other projects as well. 

You could introduce potential compatibility issues but you wouldn't know about that until you get to your build, because there's no prevalidation for that.

Reply:
> I think dotnet add package should match the VS experience for install package.

Current behavior in VS - "Clicked on Update package in PM UI. New `PackageVersion` added to csproj file and `Directory.packages.props` file was not updated". I think it is a bug because the tooling adds `PackageVersion` XML element not the `VersionOverride`.

Reply:
The VS experience is not really implemented fully yet.

Reply:
I filed an issue for the `<PackageVersion />` being added to the project: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1544117

Reply:
@jeffkl - Hope you are okay with changing the logic to add `VersionOverride` when add package is trying to update the package version. I will update the proposal shortly.

Reply:
I updated the spec to capture the summary of this conversation. No changes should be made to the `Directory.Packages.Props` file. Add `VersionOverride` attribute to the existing `PackageReference` item in .(cs/vb)proj file.

Reply:
Can I get approval on the updated spec?